contracts/deploy/02-inbox/02-arb-to-gnosis-inbox.ts

```diff
@@ -16,41 +18,70 @@ const paramsByChainId = {
   },
   HARDHAT: {
     epochPeriod: 600, // 10 minutes
+    routerAddress: ethers.constants.AddressZero,
   },
 };
 
-// TODO: use deterministic deployments
 const deployInbox: DeployFunction = async (hre: HardhatRuntimeEnvironment) => {
-  const { ethers, deployments, getNamedAccounts, getChainId, config } = hre;
+  const { ethers, deployments, getNamedAccounts, getChainId } = hre;
   const { deploy } = deployments;
-  const chainId = Number(await getChainId());
-  const { providers } = ethers;
 
-  const deployer = (await getNamedAccounts()).deployer;
-  console.log("deployer: %s", deployer);
+  // fallback to hardhat node signers on local network
+  const [namedAccounts, defaultSigner] = await Promise.all([
```

Comment:
This code does not seem equivalent to the old one. You only wanted the `defaultSigner` when there was no `deployer` on `namedAccounts`. Meanwhile, the  `getChainId`, which is always part of the happy path with the `namedAccounts`, is not resolved concurrently.

We have two options:
1. Resolve the happy path concurrently, and the fallback asynchronous. If the fallback is slow or the happy path is the most frequent outcome, I would pick this one.
2. Resolve the three behaviors concurrently otherwise.

The first one would look like:

```ts
const [namedAccounts, rawChainId] = await Promise.all([
    getNamedAccounts(),
    getChainId(),
]);

let deployer = namedAccounts.deployer;
if (!deployer) (await hre.ethers.getSingers())[0].address;

const chainId = Number(rawChainId)
```

That version resolves the happy path concurrently, falling back in a second call for the unhappy one. I would personally break `if (!deployer) (await hre.ethers.getSingers())[0].address;` line into multiple statements, as I do not like lines that perform too many operations (especially with  `await`), but that is personal preference.

The second one would look like:
```ts
const [namedAccounts, signers, rawChainId] = await Promise.all([
    getNamedAccounts(),
    hre.ethers.getSingers()
    getChainId(),
]);

const deployer = namedAccounts.deployer ?? signers[0].address
const chainId = Number(rawChainId)
```

That results in clearer code in this particular case. If the fallback operation (getting the signers) is slower, it would not be advisable.

In general, I do not think the performance difference will be critical. All operations (but getNamedAccounts(), which I do not remember the implementation details right now) looks like they will be resolving in a matter of nanoseconds. I am sharing this feedback more to have a discussion and agree on a set of shared practices.

Reply:
Thank you for your detailed analysis. option two seems better ,will fixed it .